### PR TITLE
Solidity StringXX and HashXX types change to BytesXX

### PR DESCRIPTION
**Pull request by LefterisJP**
This PR will probably require changes to our javascript stuff.

The gist of it is:

+ We replace `hash(XX*8)` and `stringXX` by `bytesXX`.
+ `bytesXX` behaves as `hash(XX*8)` in terms of convertability and operators and as `stringXX` in terms of layout in memory (alignment, etc).

+ `byte` is an alias for `bytes1`.

+ `string` is reserved for future use.

### CODE REVIEW

File: libsolidity/Types.cpp [hunk 1]
```diff
@@ -169,14 +177,10 @@ bool IntegerType::isImplicitlyConvertibleTo(Type const& _convertTo) const
...
-	{
-		StaticStringType const& convertTo = dynamic_cast<StaticStringType const&>(_convertTo);
-		return isHash() && (m_bits == convertTo.getNumBytes() * 8);
-	}
 	return _convertTo.getCategory() == getCategory() ||
-		   _convertTo.getCategory() == Category::Contract ||
-		   _convertTo.getCategory() == Category::Enum;
+        _convertTo.getCategory() == Category::Contract ||
+        _convertTo.getCategory() == Category::Enum ||
+		_convertTo.getCategory() == Category::FixedBytes;
```
Comment (LefterisJP) [inline libsolidity/Types.cpp]: seem to have an indentation issue here, will fix

File: test/SolidityEndToEndTest.cpp [hunk 2]
```diff
@@ -507,23 +507,23 @@ BOOST_AUTO_TEST_CASE(small_signed_types)
...
 							 "    large = small;\n"
 							 "  }\n"
 							 "}\n";
 	compileAndRun(sourceCode);
 	BOOST_CHECK(callContractFunction("fixed()") == encodeArgs(string("abc\0\xff__", 7)));
-	BOOST_CHECK(callContractFunction("pipeThrough(string2,bool)", string("\0\x02", 2), true) == encodeArgs(string("\0\x2", 2), true));
+	BOOST_CHECK(callContractFunction("pipeThrough(bytes2,bool)", string("\0\x02", 2), true) == encodeArgs(string("\0\x2", 2), true));
 }
-
+#
```
Comment (chriseth) [inline test/SolidityEndToEndTest.cpp]: Remove this?

File: libsolidity/LValue.cpp [hunk 3]
```diff
@@ -265,8 +266,9 @@ void StorageByteArrayElement::storeValue(Type const&, SourceLocation const&, boo
 	m_context << eth::Instruction::DUP2 << u256(0xff) << eth::Instruction::MUL
 		<< eth::Instruction::NOT << eth::Instruction::AND;
 	// stack: value ref (1<<(32-byte_number)) old_full_value_with_cleared_byte
-	m_context << eth::Instruction::SWAP1 << eth::Instruction::DUP4 << eth::Instruction::MUL
-		<< eth::Instruction::OR;
+	m_context << eth::Instruction::SWAP1;
+	m_context << (u256(1) << (256 - 8)) << eth::Instruction::DUP5 << eth::Instruction::DIV
+			  << eth::Instruction::MUL << eth::Instruction::OR;
```
Comment (LianaHus) [inline libsolidity/LValue.cpp]: indentation?

File: libsolidity/Token.h [hunk 4]
```diff
@@ -252,77 +252,44 @@ namespace solidity
...
+	K(Bytes4, "bytes4", 0)                                             \
+	K(Bytes5, "bytes5", 0)                                             \
+	K(Bytes6, "bytes6", 0)                                             \
+	K(Bytes7, "bytes7", 0)                                             \
+	K(Bytes8, "bytes8", 0)                                             \
+	K(Bytes9, "bytes9", 0)                                             \
+	K(Bytes10, "bytes10", 0)                                             \
+	K(Bytes11, "bytes11", 0)                                             \
+	K(Bytes12, "bytes12", 0)                                             \
+	K(Bytes13, "bytes13", 0)                                           \
```
Comment (LianaHus) [inline libsolidity/Token.h]: indent

File: libsolidity/Types.cpp [hunk 5]
```diff
@@ -190,10 +194,7 @@ TypePointer IntegerType::unaryOperatorResult(Token::Value _operator) const
 	// "~" is ok for all other types
 	else if (_operator == Token::BitNot)
```
Comment (chriseth) [inline libsolidity/Types.cpp]: Can be combined with the next.

File: libsolidity/Types.cpp [hunk 6]
```diff
@@ -230,14 +231,11 @@ TypePointer IntegerType::binaryOperatorResult(Token::Value _operator, TypePointe
 	// All integer types can be compared
 	if (Token::isCompareOp(_operator))
 		return commonType;
-
 	// Nothing else can be done with addresses, but hashes can receive bit operators
```
Comment (chriseth) [inline libsolidity/Types.cpp]: Delete part about hashes.

File: libsolidity/Types.cpp [hunk 7]
```diff
@@ -284,7 +282,15 @@ IntegerConstantType::IntegerConstantType(Literal const& _literal)
 
 bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	TypePointer integerType = getIntegerType();
+	auto integerType = getIntegerType();
+	if (_convertTo.getCategory() == Category::FixedBytes)
+	{
+		FixedBytesType const& convertTo = dynamic_cast<FixedBytesType const&>(_convertTo);
```
Comment (chriseth) [inline libsolidity/Types.cpp]: auto?

File: libsolidity/Types.cpp [hunk 8]
```diff
@@ -284,7 +282,15 @@ IntegerConstantType::IntegerConstantType(Literal const& _literal)
...
+		if (convertTo.getNumBytes() * 8 >= integerType->getNumBits())
```
Comment (chriseth) [inline libsolidity/Types.cpp]: Single eturn

File: libsolidity/Types.cpp [hunk 9]
```diff
@@ -284,7 +282,15 @@ IntegerConstantType::IntegerConstantType(Literal const& _literal)
 
 bool IntegerConstantType::isImplicitlyConvertibleTo(Type const& _convertTo) const
 {
-	TypePointer integerType = getIntegerType();
+	auto integerType = getIntegerType();
```
Comment (chriseth) [inline libsolidity/Types.cpp]: can be empty

File: libsolidity/Types.cpp [hunk 10]
```diff
@@ -433,50 +439,74 @@ shared_ptr<IntegerType const> IntegerConstantType::getIntegerType() const
 												 : IntegerType::Modifier::Unsigned);
 }
 
-shared_ptr<StaticStringType> StaticStringType::smallestTypeForLiteral(string const& _literal)
+shared_ptr<FixedBytesType> FixedBytesType::smallestTypeForLiteral(string const& _literal)
```
Comment (chriseth) [inline libsolidity/Types.cpp]: We might need to check whether it is a string or a number here.

File: libsolidity/Types.cpp [hunk 11]
```diff
@@ -433,50 +439,74 @@ shared_ptr<IntegerType const> IntegerConstantType::getIntegerType() const
...
+	FixedBytesType const& convertTo = dynamic_cast<FixedBytesType const&>(_convertTo);
 	return convertTo.m_bytes >= m_bytes;
 }
 
-bool StaticStringType::isExplicitlyConvertibleTo(Type const& _convertTo) const
+bool FixedBytesType::isExplicitlyConvertibleTo(Type const& _convertTo) const
 {
 	if (_convertTo.getCategory() == getCategory())
 		return true;
 	if (_convertTo.getCategory() == Category::Integer)
```
Comment (chriseth) [inline libsolidity/Types.cpp]: We allow explicit conversion from contract, so we should also allow to contract.

File: libsolidity/Types.cpp [hunk 12]
```diff
@@ -433,50 +439,74 @@ shared_ptr<IntegerType const> IntegerConstantType::getIntegerType() const
...
 	{
 		IntegerType const& convertTo = dynamic_cast<IntegerType const&>(_convertTo);
-		if (convertTo.isHash() && (m_bytes * 8 == convertTo.getNumBits()))
+		if (m_bytes * 8 <= convertTo.getNumBits())
```
Comment (chriseth) [inline libsolidity/Types.cpp]: We should not forget to lift this requirement later.

**Comment by LefterisJP** [comment]
rebased

File: libsolidity/Types.cpp (hunk 7 above)
Comment (LefterisJP) [inline libsolidity/Types.cpp]: It's easier than typing TypePointer :)
Will change it back. I believe that line must have gone through many changes in the commits of the PR and thus missed this detail

File: libsolidity/Types.cpp [hunk 13]
```diff
@@ -1124,7 +1154,7 @@ MagicType::MagicType(MagicType::Kind _kind):
 	case Kind::Block:
 		m_members = MemberList({{"coinbase", make_shared<IntegerType>(0, IntegerType::Modifier::Address)},
 								{"timestamp", make_shared<IntegerType>(256)},
-								{"blockhash", make_shared<FunctionType>(strings{"uint"}, strings{"hash"}, FunctionType::Location::BlockHash)},
+								{"blockhash", make_shared<FunctionType>(strings{"uint"}, strings{"bytes"}, FunctionType::Location::BlockHash)},
```
Comment (chriseth) [inline libsolidity/Types.cpp]: Should be bytes32

File: test/SolidityEndToEndTest.cpp [hunk 14]
```diff
@@ -1182,12 +1182,14 @@ BOOST_AUTO_TEST_CASE(send_ether)
 	BOOST_CHECK(callContractFunction("a(address,uint256)", address, amount) == encodeArgs(1));
 	BOOST_CHECK_EQUAL(m_state.balance(address), amount);
 }
-
+// TODO: Note that these tests should actually be
```
Comment (chriseth) [inline test/SolidityEndToEndTest.cpp]: No-one will see this comment when that conversion is implemented. Is it really not there?
Please rather add a story to PT.

File: test/SolidityNameAndTypeResolution.cpp [hunk 15]
```diff
@@ -1287,6 +1288,121 @@ BOOST_AUTO_TEST_CASE(storage_variable_initialization_with_incorrect_type_string)
...
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt200) == *make_shared<IntegerType>(200, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt208) == *make_shared<IntegerType>(208, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt216) == *make_shared<IntegerType>(216, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt224) == *make_shared<IntegerType>(224, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt232) == *make_shared<IntegerType>(232, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt240) == *make_shared<IntegerType>(240, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt248) == *make_shared<IntegerType>(248, IntegerType::Modifier::Unsigned));
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::UInt256) == *make_shared<IntegerType>(256, IntegerType::Modifier::Unsigned));
+
+	BOOST_CHECK(*Type::fromElementaryTypeName(Token::Bytes0) == *make_shared<FixedBytesType>(0));
```
Comment (chriseth) [inline test/SolidityNameAndTypeResolution.cpp]: What about `Token::Byte`?

File: libsolidity/Types.cpp (hunk 10 above)
Comment (LefterisJP) [inline libsolidity/Types.cpp]: only used for strings as we checked together

File: libsolidity/Types.cpp (hunk 12 above)
Comment (LefterisJP) [inline libsolidity/Types.cpp]: Will make a story for this

**Comment by chriseth** [comment]
Can be merged as soon as it is clear that changes in ethereum.js are made at short notice.

**Pull request merged**